Engineering/Backend/Bloom filter.md

```diff
@@ -0,0 +1,146 @@
+---
+tags: engineering, backend, architecture
+author: Hoang Nguyen(Harry)
+github_id: nnhuyhoang
+date: 2024-03-21
+---
+
+## What is Bloom Filter?
+A Bloom Filter is a probabilistic data structure used for testing whether an element is a member of a set or not. It's space-efficient compared to other data structures like hash tables, but it may give false positives (indicating that an element is in the set when it's not) and never gives false negatives (indicating that an element is not in the set when it actually is not).
+
+![](bloom_filter.png)
+
+## How Bloom Filter work?
+
+``Initialization``: You start with a bit array of size `m` (usually a large prime number) initially set to all zeros, and `k` different hash functions.
+
+``Adding Elements``: When you want to add an element to the Bloom Filter, you run it through each of the `k` hash functions, each of which maps the element to one of the `m` bits in the array. You then set those bits to `1`.
+
+``Checking Membership``: To check if an element is in the Bloom Filter, you again run it through each of the `k` hash functions. If all `k` bits are set to `1` in the bit array, the element is probably in the set. If any of the bits are `0`, then the element is definitely not in the set. However, due to potential hash collisions, `false positives are possible`.
+
+## When to use?
+Imagine that you have a system which millions user, and then you have a user table which had millions of records. So everytime a new user is registered, you have to query millions records table to check username is existed or not which is very bad in performance. Instead of that, you can `Checking membership` in Bloom Filter which just have `Time complexity = O(k)` (k is number of hash function) and then response to user that username is alreay existed. The result has a small probability of false positive(username not existed but response existed) due to potential hash collion but it acceptable by notify user can register with other username.
```

Comment:
We should explain the idea of `username exists or not, which is very bad in performance.`. The bloom filter is a probabilistic data structure, it means the `bloom filter` can detect a item is NOT existing in a set.